test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ModelStateDictionaryExtensionsTest.cs

```diff
@@ -0,0 +1,336 @@
+        {
+            // Arrange
+            var modelState = new ModelStateDictionary();
+
+            // Act
+            modelState.AddModelError<TestModel>(model => (object)model.Child.Value, "Message");
+
+            // Assert
+            Assert.Equal("Child.Value", modelState.Single().Key);
+            Assert.Equal("Message", modelState.Single().Value.Errors.Single().ErrorMessage);
+        }
+
+        [Fact]
+        public void AddModelError_ForExpressionWithoutStringRepresentation_AddsExpectedMessage()
+        {
+            // Arrange
+            var modelState = new ModelStateDictionary();
+
+            // Act
+            modelState.AddModelError<TestModel>(model => model.ToString(), "Message");
+
+            // Assert
+            Assert.Equal("", modelState.Single().Key);
+            Assert.Equal("Message", modelState.Single().Value.Errors.Single().ErrorMessage);
+        }
+
+        [Fact]
+        public void AddModelError_ForSingleExpression_AddsExpectedException()
+        {
+            // Arrange
+            var exception = new Exception();
+            var modelState = new ModelStateDictionary();
+
+            // Act
+            modelState.AddModelError<TestModel>(model => model.Text, exception);
+
+            // Assert
+            Assert.Equal("Text", modelState.Single().Key);
+            Assert.Same(exception, modelState.Single().Value.Errors.Single().Exception);
+        }
+
+        [Fact]
+        public void AddModelError_ForRelationExpression_AddsExpectedException()
+        {
+            // Arrange
+            var exception = new Exception();
+            var modelState = new ModelStateDictionary();
+
+            // Act
+            modelState.AddModelError<TestModel>(model => model.Child.Text, exception);
+
+            // Assert
+            Assert.Equal("Child.Text", modelState.Single().Key);
+            Assert.Same(exception, modelState.Single().Value.Errors.Single().Exception);
+        }
+
+        [Fact]
+        public void AddModelError_ForImplicitlyCastedToObjectExpression_AddsExpectedException()
+        {
+            // Arrange
+            var exception = new Exception();
+            var modelState = new ModelStateDictionary();
+
+            // Act
+            modelState.AddModelError<TestModel>(model => model.Child.Value, exception);
+
+            // Assert
+            Assert.Equal("Child.Value", modelState.Single().Key);
+            Assert.Same(exception, modelState.Single().Value.Errors.Single().Exception);
+        }
+
+        [Fact]
+        public void AddModelError_ForExplicitlyCastedToObjectExpression_AddsExpectedException()
+        {
+            // Arrange
+            var exception = new Exception();
+            var modelState = new ModelStateDictionary();
+
+            // Act
+            modelState.AddModelError<TestModel>(model => (object)model.Child.Value, exception);
```

Comment:
Check my comment above, about automatic struct casting to object. I just wanted to explicitly add this, so that it would be visible. I can remove the cast if you think it's okay.